support/builder/s3migrate.sh

```diff
@@ -0,0 +1,311 @@
+    if [ ${#artifacts[@]} -eq 0 ]; then
+        echo ""
+        echo "No artifacts found in Artifact storage path!"
+        echo "Exiting!"
+        exit 1
+    fi
+        keyMsg
+    for hart in ${artifacts[@]}; do
+        uploadMsg
+        keyId=$(generateKey ${hart})
+        putArtifacts "${hart}" "${S3_BUCKET}" "${AWS_REGION}" "${keyId}"
+    done
+}
+
+installDeps() {
+    # We use aws-cli to handle the actual upload.
+    # but we also need jq to parse the pkg info
+
+    PDEPS=("core/aws-cli" "core/jq-static")
+
+    configMsg
+
+    for bin in "${!PDEPS[@]}"; do
+        hab pkg install -b ${PDEPS[$bin]}
+    done
+}
+
+generateKey() {
+    # Generates the s3 item key from the hartfile
+    FILEIN="${1}"
+    BASE="${FILEIN##*/}"
+
+    # Get the
+    ORIGIN="$(hab pkg info -j ${FILEIN} | jq -r .origin)"
+    PKGNAME="$(hab pkg info -j ${FILEIN} | jq -r .name)"
+    PKGVER="$(hab pkg info -j ${FILEIN} | jq -r .version)"
+    RELEASE="$(hab pkg info -j ${FILEIN} | jq -r .release)"
+    ARCH=("x86_64")
+
+    SYSTEMS=("linux" "windows")
+    for sys in "${SYSTEMS}"; do
+        if  [[ "${FILEIN}" = *"${sys}"* ]]; then
+            SYSTEM="${sys}"
+        fi
+    done
+
+    if [ -z $SYSTEM ]; then
+        return 1
+    else
+        echo "${ORIGIN}/${PKGNAME}/${PKGVER}/${RELEASE}/${ARCH}/${SYSTEM}/${BASE}"
+    fi
+}
+setBucket() {
+    # Configure the bucket name to use for the upload
+    echo ""
+    echo "==========================================================="
+    echo "Please enter a target bucket name and press [ENTER]:"
+    read bucket_name
+    # Check if bucket exists, if so create it, if not continue using existing bucket
+    if [ ${s3type} == 'minio' ]; then
+        if aws --endpoint-url ${AWS_REGION} s3 ls | grep ${bucket_name} >/dev/null; then
+            echo "Bucket: ${bucket_name} found!"
+            echo "Using specified bucket."
+            export S3_BUCKET=${bucket_name}
+        else
+            echo "Bucket: ${bucket_name} not found!"
+            echo "Creating bucket: "${bucket_name}
+            aws --endpoint-url ${AWS_REGION} s3 mb s3://${bucket_name}
+            export S3_BUCKET=${bucket_name}
+        fi
+    else
+        if aws s3api list-buckets | grep ${bucket_name} >/dev/null; then
+            echo "Bucket: ${bucket_name} found!"
+            echo "Using specified bucket."
+            export S3_BUCKET=${bucket_name}
+        else
+            echo "Bucket: ${bucket_name} not found!"
+            echo "Creating bucket: "${bucket_name}
+            aws s3api create-bucket --bucket ${bucket_name} --region ${AWS_REGION}
+            export S3_BUCKET=${bucket_name}
+        fi
+    fi
+}
+
+setRegion() {
+    # Sets the region used by the bucket
+    echo ""
+    echo "==========================================================="
+    echo "Please enter the region or the minio endpoint for your bucket and press [ENTER]:"
+    echo "(us-west-1 us-east-1 || http://localhost:9000 )"
```

Comment:
Only ask for this if the endpoint is AWS?

Reply:
Nope, we need this for on-prem as well. the region/endpoint are used with minio as well. The local copy I have tries to detect the minio instance from a supervisor ring, but still allows you to specify something custom if you've got more than one instance appropriately named.